### PR TITLE
Fix YlmSpherepack

### PR DESCRIPTION
**Pull request by markscheel**
## Proposed changes

Cleanups from Nils' code review.

### Types of changes:

- [x] Bugfix
- [ ] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- [ ] Code has documentation and unit tests
- [ ] Private member variables have a trailing underscore
- [ ] Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- [ ] Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- [ ] File lists in CMake are alphabetical
- [ ] Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- [ ] Mark objects `const` whenever possible
- [ ] Almost always `auto`, except with expression templates, i.e. `DataVector`
- [ ] All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- [ ] Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."

### CODE REVIEW

File: src/ApparentHorizons/YlmSpherepack.cpp [hunk 1]
```diff
@@ -363,26 +380,27 @@ const std::vector<double>& YlmSpherepack::theta_points() const noexcept {
...
+    memory_pool_.free(work);
+    if (err != 0) {
+      ERROR("gaqd error " << err << " in YlmSpherepack");
     }
   }
   return theta;
 }
 
 const std::vector<double>& YlmSpherepack::phi_points() const noexcept {
+  static constexpr double two_pi = 2.0 * M_PI;
```
Comment (kidder) [inline src/ApparentHorizons/YlmSpherepack.cpp]: why not two_pi_over_n_phi = 2.0 * M_PI/ n_phi_

File: src/ApparentHorizons/YlmSpherepack.cpp [hunk 2]
```diff
@@ -682,26 +674,27 @@ YlmSpherepack::InterpolationInfo YlmSpherepack::set_up_interpolation_info(
 
 /// \cond DOXYGEN_FAILS_TO_PARSE_THIS
 void YlmSpherepack::interpolate(
+    gsl::not_null<std::vector<double>*> result,
```
Comment (kidder) [inline src/ApparentHorizons/YlmSpherepack.cpp]: const

File: src/ApparentHorizons/YlmSpherepack.cpp [hunk 3]
```diff
@@ -919,5 +915,5 @@ bool operator!=(const YlmSpherepack& lhs, const YlmSpherepack& rhs) noexcept {
 
 // Explicit instantiations
 template void YlmSpherepack::interpolate_from_coefs<std::vector<double>>(
-    const std::vector<double>&, const InterpolationInfo&,
-    gsl::not_null<std::vector<double>*>, size_t, size_t) const noexcept;
+    gsl::not_null<std::vector<double>*>, const std::vector<double>&,
```
Comment (kidder) [inline src/ApparentHorizons/YlmSpherepack.cpp]: const

File: src/ApparentHorizons/YlmSpherepack.hpp [hunk 4]
```diff
@@ -253,15 +255,27 @@ class YlmSpherepack {
...
+      noexcept {
+    return storage_.quadrature_weights;
+  }
 
   /// Adds a constant (i.e. \f$f(\theta,\phi)\f$ += \f$c\f$) to the function
   /// given by the spectral coefficients, by modifying the coefficients.
-  void add_constant(gsl::not_null<DataVector*> spectral_coefs, double c) const
-      noexcept;
+  SPECTRE_ALWAYS_INLINE void add_constant(
+      gsl::not_null<DataVector*> spectral_coefs, double c) const noexcept {
```
Comment (kidder) [inline src/ApparentHorizons/YlmSpherepack.hpp]: const

File: tests/Unit/ApparentHorizons/YlmTestFunctions.cpp [hunk 5]
```diff
@@ -8,10 +8,11 @@ namespace YlmTestFunctions {
 void Y00::func(const gsl::not_null<DataVector*> u, const size_t stride,
                const size_t offset, const std::vector<double>& thetas,
                const std::vector<double>& phis) const noexcept {
-  size_t s = 0;
-  for (size_t j = 0; j < phis.size(); ++j) {
+  // Can't make inv_sqrt_4_pi constexpr because sqrt isn't constexpr.
+  static const double inv_sqrt_4_pi = 1.0 / sqrt(4.0 * M_PI);
```
Comment (kidder) [inline tests/Unit/ApparentHorizons/YlmTestFunctions.cpp]: 0.5 / sqrt(M_PI)

File: src/ApparentHorizons/YlmSpherepack.hpp [hunk 6]
```diff
@@ -250,9 +251,17 @@ class YlmSpherepack {
...
-  double definite_integral(gsl::not_null<const double*> collocation_values,
-                           size_t physical_stride = 1,
-                           size_t physical_offset = 0) const noexcept;
+  SPECTRE_ALWAYS_INLINE double definite_integral(
+      gsl::not_null<const double*> collocation_values,
+      size_t physical_stride = 1, size_t physical_offset = 0) const noexcept {
+    // clang-tidy: 'do not use pointer arithmetic'
+    return ddot_(
+        n_theta_ * n_phi_, storage_.quadrature_weights.data(), 1,
+        static_cast<const double*>(collocation_values) + physical_offset,
```
Comment (kidder) [inline src/ApparentHorizons/YlmSpherepack.hpp]: can't you get the pointer from a gsl::not_null?

**Review by kidder — CHANGES_REQUESTED** [review]

File: src/ApparentHorizons/YlmSpherepack.cpp [hunk 7]
```diff
@@ -682,26 +665,27 @@ YlmSpherepack::InterpolationInfo YlmSpherepack::set_up_interpolation_info(
...
 /// \endcond
 
 template <typename T>
 void YlmSpherepack::interpolate_from_coefs(
-    const T& spectral_coefs, const InterpolationInfo& interpolation_info,
-    gsl::not_null<std::vector<double>*> result, const size_t spectral_stride,
-    const size_t spectral_offset) const noexcept {
+    const gsl::not_null<std::vector<double>*> result, const T& spectral_coefs,
+    const InterpolationInfo& interpolation_info,
+
```
Comment (wthrowe) [inline src/ApparentHorizons/YlmSpherepack.cpp]: Oops?

File: src/ApparentHorizons/YlmSpherepack.hpp [hunk 8]
```diff
@@ -246,38 +251,59 @@ class YlmSpherepack {
...
+  SPECTRE_ALWAYS_INLINE double definite_integral(
+      gsl::not_null<const double*> collocation_values,
+      size_t physical_stride = 1, size_t physical_offset = 0) const noexcept {
+    // clang-tidy: 'do not use pointer arithmetic'
+    return ddot_(n_theta_ * n_phi_, storage_.quadrature_weights.data(), 1,
+                 static_cast<const double*>(collocation_values) +  // NOLINT
+                     physical_offset,                              // NOLINT
+                 physical_stride);
+    // The static_cast above is because you can't do pointer arithmetic
+    // on gsl::not_null<const double*>
```
Comment (wthrowe) [inline src/ApparentHorizons/YlmSpherepack.hpp]: Why not `collocation_values.get()`?

File: src/ApparentHorizons/YlmSpherepackHelper.cpp [hunk 9]
```diff
@@ -20,6 +20,11 @@ std::vector<double>& MemoryPool::get(size_t n_pts) noexcept {
   ERROR("Attempt to allocate more than " << num_temps_ << " temps.");
 }
 
+// We don't simply forward this to the const double* version
+// because std::vector<T>::data() might be nullptr for vectors
+// in the pool that are unallocated, for random vectors passed into
+// this function that have nothing to do with the pool, or for vectors
+// that are in the pool but have been size-zero allocated.
 void MemoryPool::free(const std::vector<double>& to_be_freed) noexcept {
   bool found = false;
```
Comment (wthrowe) [inline src/ApparentHorizons/YlmSpherepackHelper.cpp]: This bool could be removed by changing the `break` below to a `return`.

File: src/ApparentHorizons/YlmSpherepackHelper.cpp [hunk 10]
```diff
@@ -33,7 +38,9 @@ void MemoryPool::free(const std::vector<double>& to_be_freed) noexcept {
     ERROR("Attempt to free temp that was never allocated.");
   }
 }
-void MemoryPool::free(const double* to_be_freed) noexcept {
+
+/// \cond DOXYGEN_FAILS_TO_PARSE_THIS
+void MemoryPool::free(const gsl::not_null<double*> to_be_freed) noexcept {
   bool found = false;
```
Comment (wthrowe) [inline src/ApparentHorizons/YlmSpherepackHelper.cpp]: Ditto.

**Review by kidder — DISMISSED** [review]
Nils still needs to look at these

File: src/ApparentHorizons/YlmSpherepack.cpp [hunk 11]
```diff
@@ -816,77 +799,80 @@ void YlmSpherepack::fill_vector_work_arrays() const noexcept {
...
-            dworkg.data(), static_cast<int>(ldwg), &err);
-    memory_pool_.free(dworkg);
-    ASSERT(0 == err, "Error in vhsgsi");
+  auto& dworkg = memory_pool_.get(ldwg);
+  int err = 0;
+  vhsgsi_(static_cast<int>(n_theta_), static_cast<int>(n_phi_),
+          work_vector_spec_to_phys.data(), static_cast<int>(l_vhsgs),
+          dworkg.data(), static_cast<int>(ldwg), &err);
+  memory_pool_.free(dworkg);
+  if (err != 0) {
```
Comment (nilsdeppe) [inline src/ApparentHorizons/YlmSpherepack.cpp]: Move error check to before free as is done elsewhere

File: src/ApparentHorizons/YlmSpherepack.cpp [hunk 12]
```diff
@@ -273,49 +292,51 @@ YlmSpherepack::FirstDeriv YlmSpherepack::gradient_from_coefs_all_offsets(
...
           scalar_laplacian, static_cast<int>(n_theta_),
           static_cast<int>(n_phi_), a, b, static_cast<int>(l1),
-          static_cast<int>(n_theta_), work_spec_to_phys.data(),
-          static_cast<int>(work_spec_to_phys.size()), work.data(),
+          static_cast<int>(n_theta_), work_scalar_spec_to_phys.data(),
+          static_cast<int>(work_scalar_spec_to_phys.size()), work.data(),
           static_cast<int>(work_size), &err);
   memory_pool_.free(work);
-  ASSERT(0 == err, "Slapgs failure");
+  if (err != 0) {
```
Comment (nilsdeppe) [inline src/ApparentHorizons/YlmSpherepack.cpp]: move in front of free like elsewhere

File: src/ApparentHorizons/YlmSpherepack.cpp [hunk 13]
```diff
@@ -816,77 +799,80 @@ void YlmSpherepack::fill_vector_work_arrays() const noexcept {
...
     auto& work0 = memory_pool_.get(n_theta_);
     auto& work1 = memory_pool_.get(n_theta_ + 1);
     int err = 0;
     gaqd_(static_cast<int>(n_theta_), work0.data(), weights.data(),
           work1.data(), static_cast<int>(n_theta_ + 1), &err);
     memory_pool_.free(work0);
     memory_pool_.free(work1);
-    ASSERT(0 == err, "Failure in gaqd");
-  }
+    if (err != 0) {
```
Comment (nilsdeppe) [inline src/ApparentHorizons/YlmSpherepack.cpp]: we should annotate all of the error checks with `if (UNLIKELY(err != 0))` to tell the compiler that we do not expect this to evaluate true very frequently (in practice more than once per simulation per core). The macro is defined in `Utilities/Gsl.hpp`

File: src/ApparentHorizons/YlmSpherepack.hpp [hunk 14]
```diff
@@ -164,14 +168,15 @@ class YlmSpherepack {
   /// 3-dimensional arrays (I1 x S2 topology), and the gradient is
   /// done for all 'radial' points at once by internally looping
   /// over all values of the offset from zero to `stride`-1.
-  void gradient_all_offsets(gsl::not_null<const double*> collocation_values,
-                            const std::array<double*, 2>& df,
+  void gradient_all_offsets(const std::array<double*, 2>& df,
+                            gsl::not_null<const double*> collocation_values,
                             size_t stride = 1) const noexcept;
 
   void gradient_from_coefs_all_offsets(
```
Comment (nilsdeppe) [inline src/ApparentHorizons/YlmSpherepack.hpp]: `SPECTRE_ALWAYS_INLINE`

File: src/ApparentHorizons/YlmSpherepack.hpp [hunk 15]
```diff
@@ -43,10 +44,12 @@ class YlmSpherepack {
   /// collocation points and spectral coefficients for a given l_max
   /// and m_max.  Useful for allocating space without having to create
   /// a YlmSpherepack.
-  static size_t physical_size(const size_t l_max, const size_t m_max) noexcept {
+  static constexpr size_t physical_size(const size_t l_max,
```
Comment (nilsdeppe) [inline src/ApparentHorizons/YlmSpherepack.hpp]: mark both `SPECTRE_ALWAYS_INLINE`

File: tests/Unit/ApparentHorizons/YlmTestFunctions.hpp [hunk 16]
```diff
@@ -5,6 +5,9 @@
 
 #include "ApparentHorizons/YlmSpherepack.hpp"
```
Comment (nilsdeppe) [inline tests/Unit/ApparentHorizons/YlmTestFunctions.hpp]: No blank line (this might be because of my extraneous space in my previous comment)

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
Just a few minor things, thanks for taking care of all this!

**Review by nilsdeppe — DISMISSED** [review]
Let's squash these into one commit named something like "Address late comments from PR for YlmSpherepack"

**Review by kidder — APPROVED** [review]

**Review by nilsdeppe — APPROVED** [review]

**Pull request merged**